sway: 1.0 -> 1.1.1

Synthetica9 opened the pull request:
Should be ready for merge!

~~Compiles now, thanks to a new upstream release. Still waiting for a resolution of #60434.~~

~~Currently does't compile, still investigating that.~~

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
  - [x] Runs from within another sway session (in windowed mode)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Synthetica9 committed:
wlroots: 0.5 -> 0.6

Synthetica9 committed:
sway: 1.0 -> 1.1.1

Synthetica9 commented:
cc @primeos

Synthetica9 commented:
<details><summary>Relevant output</summary>

```
[76/271] Compiling C object 'sway/345c74e@@sway@exe/desktop_output.c.o'.
FAILED: sway/345c74e@@sway@exe/desktop_output.c.o
/nix/store/ds1prvgw3i3ic8c7axyrw4lwm3d0gqab-gcc-wrapper-7.4.0/bin/cc -Isway/345c74e@@sway@exe -Isway -I../sway -Iinclude -I../include -I/nix/store/chq3xpz7r54c9lh3xf8y330zsj1w4aw3-cairo-1.16.0-dev/include/cairo -I/nix/store/z6wjrmaflgkjj3ckzqc7d0ap79rhbi1j-freetype-2.10.0-dev/include/freetype2 -I/nix/store/z6wjrmaflgkjj3ckzqc7d0ap79rhbi1j-freetype-2.10.0-dev/include -I/nix/store/g2pjsgklp091m55gvvj721p33yljzzrd-json-c-0.13.1-dev/include/json-c -I/nix/store/9sni6rbv28hdfw7h0qqyrx1chswq1srg-libevdev-1.6.0/include/libevdev-1.0/ -I/nix/store/42yvfxq74vfxd2643x4fndpq7ndiirh7-libinput-1.13.2-dev/include -I/nix/store/1s97ja307db2dv5l1qvy4xzyw9nfmkb7-pango-1.43.0-dev/include/pango-1.0 -I/nix/store/0y6pfd81k1kxbrz1nnqjnir7vihl29ry-glib-2.60.2-dev/include -I/nix/store/0y6pfd81k1kxbrz1nnqjnir7vihl29ry-glib-2.60.2-dev/include/glib-2.0 -I/nix/store/aj9v5zj7v6i6rwhw6s1c60146ljspyb5-glib-2.60.2/lib/glib-2.0/include -I/nix/store/vhbm21ns6drmg9xgcpaqlfw88gxhw2rg-pcre-8.43-dev/include -I/nix/store/b0cahknmmfrlr1pd3mnmldy98v4xyh6l-pixman-0.38.4/include/pixman-1 -I/nix/store/r0qby1xczlyiglkph4gdwdj0sgp6i9n7-wayland-1.17.0/include -I/nix/store/z82mmy9l5zwmmvfj9ali6iplbw007lgy-wlroots-0.6.0/include -I/nix/store/rcmsp0p5m6cqdsivq5hx7j1dfgaqpmpv-libxkbcommon-0.8.0-dev/include -I/nix/store/h8aqv4ckyj99bbikd5rzn37n27sxf52i-libxcb-1.13.1-dev/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Werror -std=c11 -O3 '-DSWAY_SRC_DIR="/build/source"' -DWL_HIDE_DEPRECATED -DWLR_USE_UNSTABLE -Wno-unused-parameter -Wno-unused-result -Wundef -Wvla '-DSYSCONFDIR="/etc"' '-DNIX_SYSCONFDIR="//nix/store/13qdx1yixy4xak3klv60pkll90i2zz41-sway-1.1/etc"' '-DSWAY_VERSION="1.0"'  -MD -MQ 'sway/345c74e@@sway@exe/desktop_output.c.o' -MF 'sway/345c74e@@sway@exe/desktop_output.c.o.d' -o 'sway/345c74e@@sway@exe/desktop_output.c.o' -c ../sway/desktop/output.c
../sway/desktop/output.c: In function 'update_output_manager_config':
../sway/desktop/output.c:512:3: error: implicit declaration of function 'wlr_output_configuration_v1_create'; did you mean 'wlr_output_cursor_create'? [-Werror=implicit-function-declaration]
   wlr_output_configuration_v1_create();
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   wlr_output_cursor_create
../sway/desktop/output.c:512:3: error: initialization makes pointer from integer without a cast [-Werror=int-conversion]
../sway/desktop/output.c:520:4: error: implicit declaration of function 'wlr_output_configuration_head_v1_create'; did you mean 'wlr_output_cursor_create'? [-Werror=implicit-function-declaration]
    wlr_output_configuration_head_v1_create(config, output->wlr_output);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    wlr_output_cursor_create
../sway/desktop/output.c:520:4: error: initialization makes pointer from integer without a cast [-Werror=int-conversion]
../sway/desktop/output.c:524:14: error: dereferencing pointer to incomplete type 'struct wlr_output_configuration_head_v1'
   config_head->state.enabled = output->enabled;
              ^~
../sway/desktop/output.c:531:2: error: implicit declaration of function 'wlr_output_manager_v1_set_configuration'; did you mean 'wlr_output_transformed_resolution'? [-Werror=implicit-function-declaration]
  wlr_output_manager_v1_set_configuration(server->output_manager_v1, config);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  wlr_output_transformed_resolution
../sway/desktop/output.c:531:50: error: 'struct sway_server' has no member named 'output_manager_v1'; did you mean 'idle_inhibit_manager_v1'?
  wlr_output_manager_v1_set_configuration(server->output_manager_v1, config);
                                                  ^~~~~~~~~~~~~~~~~
                                                  idle_inhibit_manager_v1
cc1: all warnings being treated as errors
```

</details>

alexarice commented:
https://github.com/NixOS/nixpkgs/issues/60434 
Are you aware of this discussion?

Synthetica9 commented:
@alexarice 	No, I was not, thank you for bringing it to my attention.

alexarice commented:
Doesn't help with trying to compile sway though

Synthetica9 commented:
(I asked on IRC, and it seems to be an upstream thing)

emersion commented:
Sorry for the mess. 1.1 is broken, we'll release 1.1.1 shortly to fix this.

alexarice commented on pkgs/applications/window-managers/sway/default.nix:
```diff
@@ -34,7 +27,7 @@ stdenv.mkDerivation rec {
     wayland libxkbcommon pcre json_c dbus libevdev
     pango cairo libinput libcap pam gdk_pixbuf
-    wlroots wayland-protocols
+    wlroots wayland-protocols swaybg
```
Have you managed to test this as it was my understanding that this would not be sufficient to get sway to find swaybg

Synthetica9 replied:
The following works, but it feels like a rushed solution to #60434.

```diff
diff --git a/nixos/modules/programs/sway.nix b/nixos/modules/programs/sway.nix
index b4f03151cdc..f92d09a7ef4 100644
--- a/nixos/modules/programs/sway.nix
+++ b/nixos/modules/programs/sway.nix
@@ -55,7 +55,7 @@ in {
     extraPackages = mkOption {
       type = with types; listOf package;
       default = with pkgs; [
-        swaylock swayidle
+        swaylock swayidle swaybg
         xwayland rxvt_unicode dmenu
       ];
       defaultText = literalExample ''
diff --git a/pkgs/applications/window-managers/sway/default.nix b/pkgs/applications/window-managers/sway/default.nix
index e2f9786bf4e..a6dd89fd90d 100644
--- a/pkgs/applications/window-managers/sway/default.nix
+++ b/pkgs/applications/window-managers/sway/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchFromGitHub, fetchpatch
+{ stdenv, fetchFromGitHub, fetchpatch, makeWrapper
 , meson, ninja
 , pkgconfig, scdoc
 , wayland, libxkbcommon, pcre, json_c, dbus, libevdev
@@ -27,7 +27,7 @@ stdenv.mkDerivation rec {
   buildInputs = [
     wayland libxkbcommon pcre json_c dbus libevdev
     pango cairo libinput libcap pam gdk_pixbuf
-    wlroots wayland-protocols swaybg
+    wlroots wayland-protocols swaybg makeWrapper
   ];
 
   enableParallelBuilding = true;
@@ -37,6 +37,10 @@ stdenv.mkDerivation rec {
     "-Dtray=enabled" "-Dman-pages=enabled"
   ];
 
+  postInstall = ''
+    wrapProgram $out/bin/sway --prefix PATH : "${swaybg}/bin"
+  '';
+
   meta = with stdenv.lib; {
     description = "i3-compatible tiling Wayland compositor";
     homepage    = https://swaywm.org;
```

alexarice replied:
I think this is essentially option 1 from #60434

Synthetica9 replied:
Yes, you're right, I included that by accident.

alexarice replied:
There also seems to be a swaybg_command argument for sway though I can't find much documentation on it

Synthetica9 replied:
Maybe add a `programs.sway.swaybg_command`, that defaults to `"${swaybg}/bin/swaybg"`?

alexarice replied:
and add this as an argument in the existing sway wrapper?

Synthetica9 replied:
Yes

primeos replied:
A quick solution for the meantime could be to patch Sway with another default for `swaybg`.

I.e. use `"${swaybg}/bin/swaybg` here:
https://github.com/swaywm/sway/blob/1.1.1/sway/config.c#L286

That would work with both Nix and NixOS and we could get away without changing the module for now, until we find a proper solution (there are also other packages like `swaylock`, `swayidle`, ... that should be handled in the same way as `swaybg` IMO).

alexarice replied:
What reasoning is there for patching sway when it already provides a way to pass in a program for the background?

primeos replied:
> What reasoning is there for patching sway when it already provides a way to pass in a program for the background?

We only modify the default value, which we can do in the source code, but we cannot modify the configuration files of existing users, if they've copied them to $HOME, etc.

> What reasoning is there for patching sway when it already provides a way to pass in a program for the background?

Update: But unfortunately something like this doesn't seem to exist (AFAIK).

I also forgot about `makeWrapper` so let's probably go with:
```
postInstall = ''
  wrapProgram $out/bin/sway --prefix PATH : "${swaybg}/bin"
'';
```
to avoid yet another patch for now.

alexarice replied:
> Update: But unfortunately something like this doesn't seem to exist (AFAIK).

Doesn't https://github.com/swaywm/sway/blob/master/sway/commands/swaybg_command.c pull in a custom command from the command line or am I just confused

alexarice replied:
I think I was confused, however if upstream sway doesn't provide a way to set the background program then surely we shouldn't either

primeos commented on pkgs/applications/window-managers/sway/default.nix:
```diff
@@ -34,7 +27,7 @@ stdenv.mkDerivation rec {
     wayland libxkbcommon pcre json_c dbus libevdev
     pango cairo libinput libcap pam gdk_pixbuf
-    wlroots wayland-protocols
+    wlroots wayland-protocols swaybg
```
Was there any reason to add `swaybg` to `buildInputs`? I haven't checked this for 1.1 but it wasn't required for 1.1-rc3 and doesn't make a difference.

Synthetica9 replied:
I read this:

> Unlike the other sway clients which have been split into separate projects, however, swaybg is a required dependency of sway.

In the release notes, and assumed that was actually the case, but turns out that's not the case. It builds without it.

primeos reviewed (changes requested):
Also: Please use `nix-review` to verify that the wlroots update doesn't break other builds (at least Waybar depends on wlroots as well).

And we have to find a solution for `swaybg` before we merge this.

Synthetica9 commented:
I tested waybar with this pr, and it works as expected.

Synthetica9 commented:
I included swaybg as a wrapper. I think this means this PR is ready for primetime!

alexarice commented on pkgs/applications/window-managers/sway/default.nix:
```diff
@@ -34,7 +27,7 @@ stdenv.mkDerivation rec {
     wayland libxkbcommon pcre json_c dbus libevdev
     pango cairo libinput libcap pam gdk_pixbuf
-    wlroots wayland-protocols
+    wlroots wayland-protocols swaybg makeWrapper
```
Unless I'm mistaken the swaybg here is still unnecessary

alexarice replied:
I believe `makeWrapper` should also be a native build input

Synthetica9 replied:
No, because it used in making the wrapper.

alexarice replied:
you pass it in with `"${swaybg}/bin"` though

Synthetica9 replied:
Oh, yeah, you're right.

alexarice commented on pkgs/applications/window-managers/sway/default.nix:
```diff
@@ -1,30 +1,23 @@
-{ stdenv, fetchFromGitHub, fetchpatch
+{ stdenv, fetchFromGitHub, fetchpatch, makeWrapper
```
fetchpatch is no longer needed

alexarice commented on pkgs/development/libraries/wlroots/default.nix:
```diff
@@ -6,15 +6,15 @@
 let
   pname = "wlroots";
-  version = "0.5.0";
+  version = "0.6.0";
```
I realise you didn't write this but is there any reason pname and version can't go inside the derivation and the `name = "${pname}-${version}"; line removed

Synthetica9 replied:
No, I don't think there is. I'll try it and see if anything breaks.

Synthetica9 replied:
(It's probably legacy though, because automatic pname-version is pretty new)

primeos reviewed (approved):
The change looks good now, thanks :)
I'll give it a quick try to double check that we didn't miss anything.

@alexarice thanks for the awesome review :)

primeos reviewed (changes requested):
Unfortunately the version wasn't updated in `meson.build` - could you include something like this (either with sed or via a real patch)?:
```
postPatch = ''
      sed -iE "s/version: '1.0'/version: '${version}'/" meson.build
    '';
```
I tested that successfully with 1.1-rc3: https://github.com/primeos/nixpkgs-primeos/blob/master/default.nix

(Note: The example only works for version 1.0 but that is on purpose as there are also other version numbers in there that we do not want to change by accident if we use regular expressions.)

**Edit**: And while we're at it, we could also consider enabling the RDP backend:
https://github.com/primeos/nixpkgs-primeos/blob/master/default.nix#L16

Synthetica9 commented:
> Edit: And while we're at it, we could also consider enabling the RDP backend:
primeos/nixpkgs-primeos:default.nix@master#L16 

Always? (I'm not saying that's a bad idea, I'm just wondering if there are any downsides)

Synthetica9 committed:
sway: manually update version in meson.build

This wasn't done upstream. Adapted from:

https://github.com/primeos/nixpkgs-primeos/blob/5956e993dd320183f943e72fd6ec200232900249/default.nix#L32-L34

Co-Authored-By: Michael Weiss <dev.primeos@gmail.com>

primeos commented:
> Always? (I'm not saying that's a bad idea, I'm just wondering if there are any downsides)

I was also wondering if this is a good idea or not. An obvious downside is that we introduce an additional dependency and increase the size, but that shouldn't make a significant difference (and we also enable all other features by default).
But we can skip it for now as there are also a few open bug reports and I didn't test it yet.

primeos reviewed (approved):
Should be good to go :)
